### PR TITLE
Add base property to the MultiArray

### PR DESCRIPTION
**Pull request by Idutvuker**

### CODE REVIEW

File: multik-api/src/test/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/SliceTest.kt [hunk 1]
```diff
@@ -53,4 +54,35 @@ class SliceTest {
         assertEquals(mk.ndarray(mk[mk[7]]), a[0..1, 1, 1, 1..2])
         assertEquals(mk.ndarrayOf(6), a[0..1, 1, 1, 0])
     }
+
+    @Test
+    fun testBase() {
+        val a = mk.ndarrayOf(0, 1, 2, 3, 4, 5)
+        val b = a[1..5] as NDArray
+        val c = a[1..3] as NDArray
```
Comment (devcrocod) [inline multik-api/src/test/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/SliceTest.kt:62]: Why is `as` here?

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt [hunk 2]
```diff
@@ -256,7 +258,7 @@ public class NDArray<T, D : Dimension> constructor(
         if (this.dim.d == -1) throw Exception("Array dimension is undefined")
         if (this.dim.d > 4) return this as NDArray<T, DN>
 
-        return NDArray(this.data, this.offset, this.shape, this.strides, this.dtype, DN(this.dim.d))
+        return NDArray(this.data, this.offset, this.shape, this.strides, this.dtype, DN(this.dim.d), base = base ?: this)
```
Comment (devcrocod) [inline multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt:261]: Should pass "base" in this function?

**Review by devcrocod — APPROVED** [review]
nice

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt (hunk 2 above)
Comment (Idutvuker) [reply multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt:261]: I think so.
Newly returned array has the same MemoryView as a current one, so we should store the base array.

File: multik-api/src/test/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/SliceTest.kt (hunk 1 above)
Comment (Idutvuker) [reply multik-api/src/test/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/SliceTest.kt]: You are right, this cast in redundant.
Removed it in a0e6eb3

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt (hunk 2 above)
Comment (devcrocod) [reply multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt:261]: But it also has the same shape, strides, offset; except dim

**Comment by Idutvuker** [comment]
Add dtype propery to MemoryView

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt [hunk 3]
```diff
@@ -25,7 +25,8 @@ public class NDArray<T, D : Dimension> constructor(
     public override val shape: IntArray,
     public override val strides: IntArray = computeStrides(shape),
     public override val dtype: DataType,
```
Comment (devcrocod) [inline multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt:27]: Why dtype remained in NDArray constructor?

**Review by devcrocod — CHANGES_REQUESTED** [review]

File: multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt (hunk 3 above)
Comment (Idutvuker) [reply multik-api/src/main/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/NDArray.kt]: Removed `dtype` parameter in d2358d7

**Review by devcrocod — APPROVED** [review]

**Pull request merged**